### PR TITLE
[Khang Hou] iP

### PR DESCRIPTION
**Pull request by tankh99**
### Baron
Behold, Baron and his new features, including
1. Malicious sarcasm
2. Database storage feaatures

Additionally he can
- Remember your face and name
- Know where you live

Just watch out for this line while reading the codebase

```java
System.out.println("If you see this, you die");
```

Future enhancements include
- [ ] OpenAI integration
- [ ] Actually being able to control you like the peasant you are as it is his royal right to do so :smile:

> I want to let you know, that you're being disrespectful to a noble. Guards, send him for execution

Is just one of the quotes that Baron will NOT say to you because I haven't implemented it!

[Visit this page to get a virus](https://example.com/)

But if you ever get a virus, just write this line in your terminal `stop virus`

I also ~~don't~~ _have_ a **virus**

### CODE REVIEW

File: src/main/java/baron/Utils/DateUtils.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeParseException;
+
+/**
+ * Manages parsing and formatting of dates
+ */
+public class DateUtils {
+
+    // 03/21/1999 1600
```
Comment (jaejayrome) [inline src/main/java/baron/Utils/DateUtils.java]: I like how you have included an example DateTime format to allow other developers to have an understanding of the format required!

File: src/main/java/baron/Baron.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
+package baron;
+
+import baron.Managers.TaskManager;
+
+import java.util.Scanner;
+
+public class Baron {
+    public static void main(String[] args) {
```
Comment (jaejayrome) [inline src/main/java/baron/Baron.java]: I feel that there could be a Javadoc comment to mention how this method would be the main entrypoint of your application?

File: src/main/java/baron/Models/Deadline.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+    public Deadline(String name, LocalDateTime deadline) {
+        super(name);
+        this.deadline = deadline;
+    }
+
+    public Deadline(int id, String name, boolean isDone, LocalDateTime deadline) {
+        super(id, name, isDone);
+        this.deadline = deadline;
+    }
```
Comment (jaejayrome) [inline src/main/java/baron/Models/Deadline.java]: I personally like how you implemented method overloading but I feel that you could include Javadoc comments for constructors as well :)

File: src/main/java/baron/Models/Event.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        return event;
+    }
+
+    public String getStartDate() {
+        return this.startDate;
+    }
+
+    public void setStartDate(String startDate) {
+        this.startDate = startDate;
+    }
```
Comment (jaejayrome) [inline src/main/java/baron/Models/Event.java]: Very good adherence to the coding style thus far! It would be better if you could include Javadoc comments for Getters and Setters as well especially for novice programmers that may be new to Java!

File: src/main/java/baron/Utils/DateUtils.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    public static final DateTimeFormatter INPUT_FORMATTER = DateTimeFormatter.ofPattern("d/M/yyyy HHmm");
+    public static final DateTimeFormatter OUTPUT_FORMATTER = DateTimeFormatter.ofPattern("d/MM/yyyy, ha");
+
+    public static LocalDateTime parseDate(String dateString) {
+        try {
+            return LocalDateTime.parse(dateString, INPUT_FORMATTER);
+        } catch (DateTimeParseException ex) {
+            throw new IllegalArgumentException("Invalid date string given. Expected format is d/M/yyyy HHmm (e.g. 21/10/1999 1600)");
+        }
```
Comment (jaejayrome) [inline src/main/java/baron/Utils/DateUtils.java]: I like how you incorporated the IllegalArgumentException class which would make it more apparent to users what the exact exception would be given your custom error message!

File: src/test/java/baron/Utils/DateUtilsTest.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+            String dateString = "1/13/1999 1600";
+            LocalDateTime actual = DateUtils.parseDate(dateString);
+        });
+
+        // Invalid year
+        assertThrows(IllegalArgumentException.class, () -> {
+            String dateString = "1/13/199 1600";
+            LocalDateTime actual = DateUtils.parseDate(dateString);
+        });
+    }
```
Comment (jaejayrome) [inline src/test/java/baron/Utils/DateUtilsTest.java]: I like how your test cases are clear and elaborate and most importantly cover all edge cases that might happen, the detailing is elaborate and concise! 👍

File: src/test/java/baron/Utils/FileUtilsTest.java [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class FileUtilsTest {
+    public static final String BASE_DIR = System.getProperty("user.dir") + "/src/test/java/baron.Database";
+
+    public FileUtilsTest() {
+
+    }
```
Comment (jaejayrome) [inline src/test/java/baron/Utils/FileUtilsTest.java]: I like how you have added an empty constructor since it may not be apparent to people who are new to Java that every class would be coupled with a default empty constructor 💯

**Review by jaejayrome — COMMENTED** [review]
LGTM! Just a few nits have been spotted but mostly looks fine 🥇

File: src/main/java/baron/Models/Event.java (hunk 4 above)
Comment (tankh99) [reply src/main/java/baron/Models/Event.java]: I think getters and setters are generally accepted to not need JavaDocs? Correct me if I'm wrong

File: src/main/java/baron/database/Database.java [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+    }
+
+    /**
+     * Appends a new line to the end of the file
+     *
+     * @param filePath Filepath to write
+     * @param line     Line to append to the end of file
+     * @return the id of the newly appended line, which is actually just lineNumber - 1
+     */
+    public static long create(Path filePath, String line) {
```
Comment (evedaktyl) [inline src/main/java/baron/database/Database.java:49]: Perhaps we could use a more appropriate name for this method, such as write/writeToFile()?

File: src/main/java/baron/utils/FileUtils.java [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * In charge of managing simple file IO such as reading and writing
+ */
+public class FileUtils {
```
Comment (evedaktyl) [inline src/main/java/baron/utils/FileUtils.java:13]: I like that you have this extraction of file IO operations.

File: src/main/java/baron/managers/TaskManager.java [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+        Task task = this.tasks.remove(i);
+        TaskType type = getTaskType(task.toString());
+        TaskDao.delete(type.getCommand(), task.getId());
+        return UiManager.delete(task, this.tasks.size());
+    }
+
+    /**
+     * Prints out a list of tasks that match the search term (fuzzy search)
+     *
+     * @param input user input
```
Comment (evedaktyl) [inline src/main/java/baron/managers/TaskManager.java:153]: Perhaps we could be consistent in starting the param descriptors with a capital letter to enhance readability.

File: src/main/java/baron/managers/UiManager.java [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+        return formatOutput(sb.toString());
+    }
+
+    /**
+     * Informs user of deleted content
+     *
+     * @param task  Task added
+     * @param count count of tasks currently in list after modification
+     */
+    public static String delete(Task task, int count) {
```
Comment (evedaktyl) [inline src/main/java/baron/managers/UiManager.java]: Could we consider naming this method as showDeleteMessage() to enhance clarity?

**Review by evedaktyl — COMMENTED** [review]
Overall code is readable and neat. Apart from a few method names that could benefit from slightly more clarity, code quality is great!
